torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -289,47 +297,60 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
-            {(int64_t)pg_->getRank(),
-             (int64_t)serializedPayload.length(),
-             (int64_t)work.message_.type()},
-            {torch::kLong})};
-
-        // ProcessGroup is not thread-safe when sending with the same tag, hence
-        // the lock
-        std::vector<std::shared_ptr<c10d::ProcessGroup::Work>> pendingSends;
-        const auto& dst = work.to_.id_;
-
-        if (work.message_.isShutdown()) {
-          pendingSends.reserve(1);
-          {
-            std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
-            pendingSends.emplace_back(
-                pg_->send(preamble, dst, dst /* channelTag */));
+        try {
+          std::stringstream ss;
+          serialize(work.message_, ss);
+          std::string serializedPayload = ss.str();
+
+          std::vector<torch::Tensor> preamble = {torch::tensor(
+              {(int64_t)pg_->getRank(),
+               (int64_t)serializedPayload.length(),
+               (int64_t)work.message_.type()},
+              {torch::kLong})};
+
+          // ProcessGroup is not thread-safe when sending with the same tag,
+          // hence the lock
+          std::vector<std::shared_ptr<c10d::ProcessGroup::Work>> pendingSends;
+          const auto& dst = work.to_.id_;
+
+          if (work.message_.isShutdown()) {
+            pendingSends.reserve(1);
+            {
+              std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
+              pendingSends.emplace_back(
+                  pg_->send(preamble, dst, dst /* channelTag */));
+            }
+          } else {
+            std::vector<torch::Tensor> payload = {torch::from_blob(
+                (void*)serializedPayload.c_str(),
+                serializedPayload.length(),
+                {torch::kChar})};
+            pendingSends.reserve(2);
+
+            sendCounts_.increment(dst);
+
+            {
+              std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
+              pendingSends.emplace_back(
+                  pg_->send(preamble, dst, dst /* channelTag */));
+              pendingSends.emplace_back(
+                  pg_->send(payload, dst, dst /* channelTag */));
+            }
           }
-        } else {
-          std::vector<torch::Tensor> payload = {torch::from_blob(
-              (void*)serializedPayload.c_str(),
-              serializedPayload.length(),
-              {torch::kChar})};
-          pendingSends.reserve(2);
-
-          sendCounts_.increment(dst);
-
-          {
-            std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
-            pendingSends.emplace_back(
-                pg_->send(preamble, dst, dst /* channelTag */));
-            pendingSends.emplace_back(
-                pg_->send(payload, dst, dst /* channelTag */));
+          for (auto& pendingSend : pendingSends) {
+            pendingSend->wait();
+          }
+        } catch (std::exception& e) {
+          if (work.message_.isRequest()) {
```

Comment:
The code that calls `enqueueSend` marks the future as completed if the message is not a request, so we only handle the exception if the message is a request. This is because we can't mark a future as completed more than once

Reply:
Yes, we currently don't have a vehicle to pass the error to the response sender. Eventually, we should:

1. retry sending responses
2. If failed a few times, pass this error to application (e.g., we could expose an API for applications to register error listeners).